### PR TITLE
BUG:reorder type check/conversion so wide_to_long handles str arg for…

### LINKED ISSUE #22468: wide_to_long mishandles string arg for `stubnames`
wide_to_long should sanitize/format input args BEFORE sanity checking. 

This bug report pertains to wide_to_long() in master/pandas/core/reshape/melt.py

#### Code Sample, a copy-pastable example if possible

```python
# This check (~line 411) needs to move below input sanitization (~line 425).
 if any(col in stubnames for col in df.columns):
        raise ValueError("stubname can't be identical to a column name")
```
#### Problem description

The intention of the sanity checking code is to check if any element of the list stubname is in the the list of column names and generate an exception (line 411). Unfortunately, this check occours before the input sanitization that ensures that stubnames is a list -- it is also permitted to be string. As it is, the sanity check sometimes looks for substrings in a string and sometimes looks for substrings in a list, depending on the arg type for stubnames. 

#### Expected Output
na

#### Output of ``pd.show_versions()``
It is in master branch on github as of 8/22/2018

### PR DESCRIPTION
**Pull request by csmcallister**
closes #22468
 
- [x] closes #xxxx
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by gfyoung** [comment]
@csmcallister 

* Great to hear you love `pandas`!
* Good start, but we're going to need a test and `whatsnew` entry as well

File: pandas/tests/reshape/test_wide_to_long.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
+import pandas as pd
+from pandas.util import testing as tm
+import pytest
+import numpy as np
+
+
+@pytest.fixture
+def buggy_wide_df():
+    #Here, we have the rare case that a column name (A) is a substring
```
Comment (jreback) [inline pandas/tests/reshape/test_wide_to_long.py]: lint issues here

File: pandas/tests/reshape/test_wide_to_long.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
```
Comment (jreback) [inline pandas/tests/reshape/test_wide_to_long.py]: can you move existing wide_to_long tests from other reshaping tests to here (in test_melt)

File: pandas/tests/reshape/test_wide_to_long.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    return wide_df
+
+
+def test_reshape_bug_fix(buggy_wide_df):
+
+    expected_long_df = pd.wide_to_long(buggy_wide_df,
+                                       stubnames=['PA'],
+                                       i = ['node_id', 'A'],
+                                       j = 'time')
+    #this used to raise an error
```
Comment (jreback) [inline pandas/tests/reshape/test_wide_to_long.py]: don't need the comment, rather put a comment on the isseue for the test

File: pandas/tests/reshape/test_wide_to_long.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    long_df = pd.wide_to_long(buggy_wide_df,
```
Comment (jreback) [inline pandas/tests/reshape/test_wide_to_long.py]: call these expected & result

File: pandas/tests/reshape/test_wide_to_long.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                              stubnames='PA',
+                              i = ['node_id', 'A'],
+                              j = 'time')
+
+    tm.assert_frame_equal(long_df, expected_long_df)
+
+def test_reshape(wide_df):
+
```
Comment (jreback) [inline pandas/tests/reshape/test_wide_to_long.py]: comment for the issue

File: pandas/tests/reshape/test_wide_to_long.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    l = pd.wide_to_long(wide_df,
+                        stubnames='ht',
+                        i=['famid', 'birth'],
+                        j='age')
+    #Going from long back to wide using unstack
+    w = l.unstack()
+    w.columns = w.columns.map('{0[0]}{0[1]}'.format)
+    w.reset_index(inplace=True)
```
Comment (jreback) [inline pandas/tests/reshape/test_wide_to_long.py]: on't use inplace, reassign

File: pandas/tests/reshape/test_wide_to_long.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    tm.assert_frame_equal(long_df, expected_long_df)
+
+def test_reshape(wide_df):
+
+    l = pd.wide_to_long(wide_df,
+                        stubnames='ht',
+                        i=['famid', 'birth'],
+                        j='age')
+    #Going from long back to wide using unstack
+    w = l.unstack()
```
Comment (jreback) [inline pandas/tests/reshape/test_wide_to_long.py]: call these expected & result

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/reshape/test_wide_to_long.py (hunk 1 above)
Comment (csmcallister) [reply pandas/tests/reshape/test_wide_to_long.py]: Thank you! I followed all the links to the pep8speaks output and made changes accordingly.

File: pandas/tests/reshape/test_wide_to_long.py (hunk 3 above)
Comment (csmcallister) [reply pandas/tests/reshape/test_wide_to_long.py]: done

File: pandas/tests/reshape/test_wide_to_long.py (hunk 2 above)
Comment (csmcallister) [reply pandas/tests/reshape/test_wide_to_long.py]: Thanks for all the help on my PR!

I made the style fixes and moved the test to the TestWideToLong class in test_melt.py. I then ran pytest locally and passed. Before updating the PR, I merged master changes from the past few days to my branch doing:
```git checkout master```
```git pull upstream master --ff-only```
```git checkout my-branch-name```
```git fetch upstream```
```git merge upstream/master```
There weren't any conflicts, but to be safe I figured I'd run test_melt.py again. That's when I got the following error:
```
(pandas-dev) Charless-MacBook-Air:reshape mcallistercs$ pytest test_melt.py 
Traceback (most recent call last):
  File "/Users/mcallistercs/virtualenvs/pandas-dev/lib/python3.6/site-packages/_pytest/config/__init__.py", line 377, in _getconftestmodules
    return self._path2confmods[path]
KeyError: local('/Users/mcallistercs/Desktop/Github/pandas-csmcallister/pandas/tests/reshape/test_melt.py')

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/mcallistercs/virtualenvs/pandas-dev/lib/python3.6/site-packages/_pytest/config/__init__.py", line 377, in _getconftestmodules
    return self._path2confmods[path]
KeyError: local('/Users/mcallistercs/Desktop/Github/pandas-csmcallister/pandas/tests/reshape')

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/mcallistercs/virtualenvs/pandas-dev/lib/python3.6/site-packages/_pytest/config/__init__.py", line 408, in _importconftest
    return self._conftestpath2mod[conftestpath]
KeyError: local('/Users/mcallistercs/Desktop/Github/pandas-csmcallister/pandas/conftest.py')

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/mcallistercs/virtualenvs/pandas-dev/lib/python3.6/site-packages/_pytest/config/__init__.py", line 414, in _importconftest
    mod = conftestpath.pyimport()
  File "/Users/mcallistercs/virtualenvs/pandas-dev/lib/python3.6/site-packages/py/_path/local.py", line 668, in pyimport
    __import__(modname)
  File "/Users/mcallistercs/virtualenvs/pandas-dev/lib/python3.6/site-packages/_pytest/assertion/rewrite.py", line 226, in load_module
    py.builtin.exec_(co, mod.__dict__)
  File "/Users/mcallistercs/Desktop/Github/pandas-csmcallister/pandas/conftest.py", line 458, in <module>
    from hypothesis import strategies as st
ModuleNotFoundError: No module named 'hypothesis'
ERROR: could not load /Users/mcallistercs/Desktop/Github/pandas-csmcallister/pandas/conftest.py
```
I then opened a new shell and pip installed hypothesis, but still received the same error when trying to test again. Any help would be greatly appreciated!

**Comment by csmcallister** [comment]
Well, this is embarrassing. Copying and pasting jacked up the indentation and angered pep8speaks. Weird since because beforehand I ran ```git diff upstream/master -u -- "*.py" | flake8 --diff``` and nothing printed out.

**Comment by jreback** [comment]
looks good. can you rebase on master and push again. ping on green.

**Review by jreback — APPROVED** [review]

**Comment by csmcallister** [comment]
Rebased on master, squashing commits into one. @jreback

**Comment by jreback** [comment]
not sue why failing, can you merge master and try again

**Comment by csmcallister** [comment]
Just merged and trying again.

**Comment by jreback** [comment]
thanks @csmcallister

**Pull request merged**